### ISSUE TITLE
"TypeError: onigBinding.UTF8ToString is not a function" after some time / parallel requests in Next.js

### ISSUE THREAD

**Issue by hlubek** [shikijs/shiki#693]
### Validations

- [X] Follow our [Code of Conduct](https://github.com/shikijs/shiki/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guide](https://github.com/shikijs/shiki/blob/main/CONTRIBUTING.md).
- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.

### Describe the bug

We experienced crashes in production when using shikijs in a Next.js 14 application in React Server Components.

Initially the rendering works, but after a few (parallel?) requests we have the following stacktrace and errors are returned in the rendering:

```
TypeError: onigBinding.UTF8ToString is not a function
    at throwLastOnigError (file:///app/node_modules/@shikijs/core/dist/index.mjs:4677:38)
    at new OnigScanner (file:///app/node_modules/@shikijs/core/dist/index.mjs:4883:13)
    at createOnigScanner (file:///app/node_modules/@shikijs/core/dist/index.mjs:5024:12)
    at Object.createOnigScanner (file:///app/node_modules/@shikijs/core/dist/index.mjs:5358:20)
    at Grammar.createOnigScanner (file:///app/node_modules/@shikijs/core/dist/textmate.mjs:2275:30)
    at new CompiledRule (file:///app/node_modules/@shikijs/core/dist/textmate.mjs:1270:32)
    at RegExpSourceList._resolveAnchors (file:///app/node_modules/@shikijs/core/dist/textmate.mjs:1260:16)
    at RegExpSourceList.compileAG (file:///app/node_modules/@shikijs/core/dist/textmate.mjs:1237:56)
    at IncludeOnlyRule.compileAG (file:///app/node_modules/@shikijs/core/dist/textmate.mjs:728:57)
    at prepareRuleSearch (file:///app/node_modules/@shikijs/core/dist/textmate.mjs:2107:30)
TypeError: onigBinding.UTF8ToString is not a function
    at throwLastOnigError (file:///app/node_modules/@shikijs/core/dist/index.mjs:4677:38)
    at new OnigScanner (file:///app/node_modules/@shikijs/core/dist/index.mjs:4883:13)
    at createOnigScanner (file:///app/node_modules/@shikijs/core/dist/index.mjs:5024:12)
    at Object.createOnigScanner (file:///app/node_modules/@shikijs/core/dist/index.mjs:5358:20)
    at Grammar.createOnigScanner (file:///app/node_modules/@shikijs/core/dist/textmate.mjs:2275:30)
    at new CompiledRule (file:///app/node_modules/@shikijs/core/dist/textmate.mjs:1270:32)
    at RegExpSourceList.compile (file:///app/node_modules/@shikijs/core/dist/textmate.mjs:1219:28)
    at RegExpSourceList.compileAG (file:///app/node_modules/@shikijs/core/dist/textmate.mjs:1225:25)
    at IncludeOnlyRule.compileAG (file:///app/node_modules/@shikijs/core/dist/textmate.mjs:728:57)
    at prepareRuleSearch (file:///app/node_modules/@shikijs/core/dist/textmate.mjs:2107:30)
[Error: An error occurred in the Server Components render. The specific message is omitted in production builds to avoid leaking sensitive details. A digest property is included on this error instance
  digest: '25876654'
}
[Error: An error occurred in the Server Components render. The specific message is omitted in production builds to avoid leaking sensitive details. A digest property is included on this error instance
  digest: '3589831450'
}
```

This is our code:

```tsx
export default async function Code({ title, code, language }: CodeProps) {
  const html = await codeToHtml({
    code,
    language,
  });

  return <div
        dangerouslySetInnerHTML={{ __html: html }}
      />
}
```

Where our syntax definition comes from https://github.com/networkteam/vscode-neos-fusion/blob/master/syntaxes/fusion.tmLanguage.json .

```tsx
import { bundledLanguages, getHighlighter } from 'shiki/bundle/web';
import fusion from '../data/fusion.tmLanguage.json';

export const codeToHtml = async ({ code, language }: { code: string; language: string }) => {
  const highlighter = await getHighlighter({
    themes: ['dark-plus'],
    langs: [...Object.keys(bundledLanguages), fusion as any],
  });

  return highlighter.codeToHtml(code, {
    lang: language,
    themes: {
      dark: 'dark-plus',
      light: 'dark-plus',
    },
    colorReplacements: {
      '#1e1e1e': '#282828',
    },
  });
};
```

We could reproduce the behaviour with different Next.js versions (>=14) and by just refreshing constantly with F5 in the browser.

Our workaround for now is to use `unsable_cache` to cache the generated highlighted HTML.

### Reproduction

https://codesandbox.io/p/devbox/silly-grass-ngrp2h

### Contributes

- [ ] I am willing to submit a PR to fix this issue
- [ ] I am willing to submit a PR with failing tests

**Comment by zhangyu1818** [comment]
Same error, but I'm using`@shikijs/rehype` in mdx, and I don't know how to solve it yet.

When I reverted to the code that I could deploy yesterday and redeployed it, it still failed. 

I'm not sure if this problem is caused by `Next.js` or `Shiki`, but I have seen this error many times in the development environment, and I have to restart the development server every time.

By the way, I don't think this title describes the error very well😂

**Comment by Neo-Zhixing** [comment]
Ran into the same problem in an application that uses `@nuxtjs/mdc` when preprocessing a large number of markdown files. I reverted shiki version as far back as `1.6.0` but that didn't make a difference.

**Comment by slax57** [comment]
@Neo-Zhixing I believe the code responsible for this (at least, according to the stacktrace) was introduced in shiki `1.0.0`, so you'd have to downgrade to `v0` to have a chance to get it working.
However in my case I couldn't do that because I'm using the rehype plugin, so can't be affirmative this will work.

**Comment by antfu** [comment]
This down to two problems

- `UTF8ToString` was missing during throwing the error, where the error is `fail to memory allocation`
	- fixed in https://github.com/shikijs/shiki/commit/0c89157fb359ed460786af5a10b36589ea40214a
- Memory leak was due to `getHighlighter` being called too many times, where the instance should be cached
	- improved the docs and added a runtime warning in https://github.com/shikijs/shiki/commit/412fe10ef185a329ca9d5e2f722751e3cd7389a5

**Comment by Neo-Zhixing** [comment]
Thanks! That was fast.

**Comment by slax57** [comment]
Thanks for the update!

For those using the `@shikijs/rehype` plugin just like me, note that you can pass a `getHighlighter` singleton instance to the plugin by following the instructions from the [Fine-grained Bundle](https://shiki.style/packages/rehype#fine-grained-bundle) doc section:

```ts
import { unified } from 'unified'
import remarkParse from 'remark-parse'
import remarkRehype from 'remark-rehype'
import rehypeStringify from 'rehype-stringify'
import rehypeShikiFromHighlighter from '@shikijs/rehype/core'

import { getHighlighterCore } from 'shiki/core'

const highlighter = await getHighlighterCore({
  themes: [
    import('shiki/themes/vitesse-light.mjs')
  ],
  langs: [
    import('shiki/langs/javascript.mjs'),
  ],
  loadWasm: import('shiki/wasm')
})

const raw = await fs.readFile('./input.md')
const file = await unified()
  .use(remarkParse)
  .use(remarkRehype)
  .use(rehypeShikiFromHighlighter, highlighter, {
    //                             ^^^^^^^^^^^ singleton instance
    themes: {
      light: 'vitesse-light',
      dark: 'vitesse-dark',
    }
  })
  .use(rehypeStringify)
  .processSync(raw) // it's also possible to process synchronously
```

Btw, maybe it would be worth mentioning this good practice in the plugin's doc page (https://shiki.style/packages/rehype) too @antfu , wdyt?

**Comment by antfu** [comment]
Yeah sure, PR welcome. I also work on #702 to make it less confusing.

**Comment by Gerrit0** [comment]
> Memory leak was due to getHighlighter being called too many times, where the instance should be cached

This raises the question... is there a way to ask Shiki to dispose of that memory?

**Comment by antfu** [comment]
Found a proper way to dispose the instance https://github.com/shikijs/shiki/pull/707, will landed in v1.8
